### PR TITLE
Destructible floors

### PR DESCRIPTION
**Pull request by BryanFriestad**

### CODE REVIEW

File: src/main/java/pro/trevor/tankgame/state/board/unit/GenericTank.java [hunk 1]
```diff
@@ -11,16 +11,20 @@
...
-        super(position, defaults);
+        super(defaults);
         this.player = player;
+        this.position = position;
     }
 
     public GenericTank(JSONObject json) {
         super(json);
         this.player = json.getString("name");
+        this.position = new Position(json.getString("position"));
```
Comment (ryan3r) [inline src/main/java/pro/trevor/tankgame/state/board/unit/GenericTank.java:27]: This code gets duplicated a lot.  Having a GenericPositionedElement might be a good idea.

File: src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java [hunk 2]
```diff
@@ -209,8 +209,19 @@ public static <T extends GenericTank> PlayerActionRule<T> SpendActionToShootWith
                 }
                 case BasicWall wall -> wall.setDurability(wall.getDurability() - 1);
                 case EmptyUnit emptyUnit -> {
-                    /* MISS */ }
-                default -> throw new Error("Unhandled tank shot onto " + u.getClass().getName());
+
+                }
+                case DestructibleFloor floor -> {
```
Comment (ryan3r) [inline src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java:214]: You should test shooting a destructible floor with a unit on it.

File: src/main/java/pro/trevor/tankgame/state/board/Board.java [hunk 3]
```diff
@@ -73,6 +73,16 @@ public Optional<IFloor> getFloor(Position p) {
         return getElementOnBoard(floorBoard, p);
     }
 
+    // Returns the unit at the position if there is one.
+    // If there is no unit at the position, then the floor is returned.
+    public Optional<IElement> getTopElement(Position p) {
```
Comment (ryan3r) [inline src/main/java/pro/trevor/tankgame/state/board/Board.java:78]: I see where the name getTopElement came from but it sounds like it's getting the Unit.  I think something like getUnitOrFloor is more descriptive.

**Review by ryan3r — CHANGES_REQUESTED** [review]

File: src/main/java/pro/trevor/tankgame/state/board/Board.java (hunk 3 above)
Comment (BryanFriestad) [reply src/main/java/pro/trevor/tankgame/state/board/Board.java:78]: Why would it sound like that? Element != Unit. I did start with that name, but opted for this one instead. Not a big deal to switch back, but I want to understand why.

File: src/main/java/pro/trevor/tankgame/state/board/unit/GenericTank.java (hunk 1 above)
Comment (BryanFriestad) [reply src/main/java/pro/trevor/tankgame/state/board/unit/GenericTank.java:27]: I've already pointed out that `IElement`, `IUnit`, `IFloor`, `IPositioned`, etc. need to be clarified. To my understanding, `IElement`s should have a reportable position, since they are "board elements". 

Additionally, I think it's detail-leaking for all of the State types/classes to know that they are reported via JSON. A translation layer that the State component doesn't depend on could be better.

Basically, I would like to defer this until we have a design in place.

File: src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java (hunk 2 above)
Comment (BryanFriestad) [reply src/main/java/pro/trevor/tankgame/rule/impl/shared/rule/PlayerRules.java:214]: ACK. The expected behaviour is that the unit takes damage first while a unit is there.

File: src/main/java/pro/trevor/tankgame/state/board/Board.java (hunk 3 above)
Comment (BryanFriestad) [reply src/main/java/pro/trevor/tankgame/state/board/Board.java:78]: Eh, I accept the other name actually, no need to respond.

**Review by ryan3r — APPROVED** [review]
